File: src/ackward/datetime/Date.cpp
#include <ackward/datetime/Date.hpp>

#include <boost/python/extract.hpp>
#include <boost/python/tuple.hpp>

// #include <ackward/core/Exceptions.hpp>
// #include <ackward/datetime/Module.hpp>
// #include <ackward/datetime/TimeDelta.hpp>

using namespace boost::python;

namespace ackward { namespace datetime
{

// Date::Date(boost::python::object obj) :
//     core::Object (obj)
// {}

// Date::Date(unsigned int year, unsigned int month, unsigned int day) 
// try :
//     core::Object (
//         module().attr("date")(year, month, day) )
// {
// }
// catch (const error_already_set&)
// {
//     core::translatePythonException();
// }

// Date Date::today() {
//     return extract<Date>(
//         module().attr("date").attr("today"));
//     // return Date(module().attr("date").attr("today")());
// }

// Date Date::fromtimestamp(double timestamp)
// {
//     try {
//         return Date(module().attr("date").attr("fromtimestamp")(timestamp));
//     } catch (const error_already_set&) {
//         core::translatePythonException();
//         throw;
//     }
// }

// Date Date::fromordinal(unsigned int ordinal)
// {
//     try {
//         return Date(module().attr("date").attr("fromordinal")(ordinal));
//     } catch (const error_already_set&) {
//         core::translatePythonException();
//         throw;
//     }
// }

// Date Date::min()
// {
//     return Date(module().attr("date").attr("min"));
// }

// Date Date::max()
// {
//     return Date(module().attr("date").attr("max"));
// }

// TimeDelta Date::resolution()
// {
//     return TimeDelta(module().attr("date").attr("resolution"));
// }

// Date Date::replace(unsigned int y,
//                    unsigned int m,
//                    unsigned int d) const
// {
//     if (y == 0) y = year();
//     if (m == 0) m = month();
//     if (d == 0) d = day();
    
//     return Date(
//         obj().attr("replace")(y, m, d));
// }

// tm Date::timetuple() const
// {
//     return extract<tm>(
//         obj().attr("timetuple")());
// }

// int Date::weekday() const
// {
//     return extract<int>(obj().attr("weekday")());
// }

// int Date::isoweekday() const
// {
//     return extract<int>(obj().attr("isoweekday")());
// }

boost::tuple<int, int, int> Date::isocalendar() const
{
    tuple t = extract<tuple>(_isocalendar());
    return boost::make_tuple(
        extract<int>(t[0])(),
        extract<int>(t[1])(),
        extract<int>(t[2])());
}

// std::string Date::isoformat() const
// {
//     return extract<std::string>(obj().attr("isoformat")());
// }

// std::string Date::ctime() const
// {
//     return extract<std::string>(obj().attr("ctime")());
// }

// std::wstring Date::strftime(const std::wstring& fmt) const
// {
//     return extract<std::wstring>(obj().attr("strftime")(fmt));
// }

std::ostream& operator<<(std::ostream& os, const Date& d)
{
    os << d.isoformat();
    return os;
}

bool operator==(const Date& lhs, const Date& rhs)
{
    return (lhs.year() == rhs.year() &&
            lhs.month() == rhs.month() &&
            lhs.day() == rhs.day());
}

}}

File: src/ackward/datetime/Date.py
from ackward import (Class,
                     class_method,
                     ClassProperty,
                     Constructor,
                     InlineFunction,
                     method,
                     Method,
                     Namespace,
                     Property,
                     TranslationUnit)

equality_operator='''
bool operator==(const Date& d) const
{ return obj() == d.obj(); }
'''

stream_operator='''
inline std::ostream& operator<<(std::ostream& os, const Date& d) {
  os << d.isoformat();
  return os;
}
'''

replace_method='''
Date replace(unsigned int y, unsigned int m, unsigned int d) const {
  y = (y == 0) ? year : y;
  m = (m == 0) ? month : m;
  d = (d == 0) ? day : d;
  return _replace(y, m, d);
}
'''

def tunit():
    return TranslationUnit(
        forward_declarations=[
            ('ackward', 'datetime', 'class TimeDelta'),
            ],
        header_includes=[
            ('boost', 'tuple', 'tuple.hpp')
            ],
        impl_includes=[
            ('ackward', 'datetime', 'Date.hpp'),
            ('ackward', 'datetime', 'TimeDelta.hpp'),
            ])

def date_class(parent):
    cls = Class(name='Date',
                wrapped_class='datetime.date',
                parent=parent)

    Constructor(
        signature=[('unsigned int', 'year'),
                   ('unsigned int', 'month'),
                   ('unsigned int', 'day')],
        parent=cls)

    class_method('Date today()', parent=cls)
    class_method('Date fromtimestamp(double timestamp)', parent=cls)
    class_method('Date fromordinal(unsigned int ordinal)', parent=cls)

    ClassProperty(
        name='min',
        type='Date',
        read_only=True,
        parent=cls)

    ClassProperty(
        name='max',
        type='Date',
        read_only=True,
        parent=cls)

    ClassProperty(
        name='resolution',
        type='TimeDelta',
        read_only=True,
        parent=cls)

    Property(
        name='year',
        type='unsigned int',
        read_only=True,
        parent=cls)

    Property(
        name='month',
        type='unsigned int',
        read_only=True,
        parent=cls)

    Property(
        name='day',
        type='unsigned int',
        read_only=True,
        parent=cls)

    Method(
        name='_replace',
        python_name='replace',
        return_type='Date',
        signature=[('unsigned int', 'year'),
                   ('unsigned int', 'month'),
                   ('unsigned int', 'day')],
        const=True,
        parent=cls)

    InlineFunction(
        code=replace_method,
        parent=cls)

    for m in [
        'tm timetuple() const',
        'unsigned int toordinal() const',
        'int weekday() const',
        'int isoweekday() const',
        'boost::tuple<int, int, int> isocalendar() const',
        'std::string isoformat() const',
        'std::string ctime() const',
        'std::wstring strftime(std::wstring fmt) const']:
        method(m, parent=cls)

    InlineFunction(
        code=equality_operator,
        parent=cls)

def definition(env):
    t = tunit()

    n = Namespace('ackward', 'datetime', parent=t)
    date_class(parent=n)

    InlineFunction(
        code=stream_operator,
        parent=n)

    return t
